src/lib.rs

```diff
@@ -27,102 +28,1070 @@ pub extern "C" fn wapc_init() {
+    let mut security_context = security_context_option.unwrap_or_default();
+    match validation_request
+        .settings
+        .supplemental_groups
+        .rule
+        .as_str()
+    {
+        "MustRunAs" => {
+            if let Some(group_ids) = security_context.supplemental_groups {
+                for group_id in group_ids {
+                    if !is_valid_id(
+                        group_id,
+                        &validation_request.settings.supplemental_groups.ranges,
+                    ) {
+                        return Err("Group ID is outside defined ranges".to_string());
+                    }
+                }
+            } else {
+                let default_group_id =
+                    validation_request.settings.supplemental_groups.ranges[0].min;
+                security_context.supplemental_groups = Some(vec![default_group_id]);
+                return Ok(Some(security_context));
+            }
+        }
+        "MayRunAs" => {
+            if let Some(group_ids) = security_context.supplemental_groups {
+                for group_id in group_ids {
+                    if !is_valid_id(
+                        group_id,
+                        &validation_request.settings.supplemental_groups.ranges,
+                    ) {
+                        return Err("Group ID is outside defined ranges".to_string());
+                    }
+                }
+            }
+        }
+        &_ => (),
+    }
+    Ok(None)
+}
+
 fn validate(payload: &[u8]) -> CallResult {
+    info!(LOG_DRAIN, "starting validation");
     let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
+    let mut pod: apicore::Pod =
+        serde_json::from_value(validation_request.request.object.clone())
+            .map_err(|e| anyhow!("Error deserializing Pod specification: {:?}", e))?;
+    let mut mutated: bool = false;
 
-    info!(LOG_DRAIN, "starting validation");
+    if pod.spec.is_none() {
+        return kubewarden::accept_request();
+    }
+    let mut pod_spec = pod.spec.unwrap();
 
-    // TODO: you can unmarshal any Kubernetes API type you are interested in
-    match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
-        Ok(pod) => {
-            // TODO: your logic goes here
-            if pod.metadata.name == Some("invalid-pod-name".to_string()) {
-                let pod_name = pod.metadata.name.unwrap();
-                info!(
-                    LOG_DRAIN,
-                    "rejecting pod";
-                    "pod_name" => &pod_name
-                );
-                kubewarden::reject_request(
-                    Some(format!("pod name {} is not accepted", &pod_name)),
-                    None,
-                )
-            } else {
-                info!(LOG_DRAIN, "accepting resource");
-                kubewarden::accept_request()
+    for mut container in pod_spec.containers.iter_mut() {
```

Comment:
We have to check the init containers too, they can have a security context. We can reuse the code below by making it abstract.